(docs): add quarkus java example (#1047)
knative/docs: community/samples/serving/helloworld-java-quarkus/README.md

# Hello World - Quarkus Java sample

> A simple JAX-RS REST API application written in Java using [QuarkusIO](https://quarkus.io/).

Reviewer: Consider applying these suggestions to add clarity and target the subject (as "It" in the second sentence might refer to QuarkusIO).

It uses an ENV variable `MESSAGE_PREFIX`(defaults to value `Hello`) and prints a message of format
`${MESSAGE_PREFIX} Knative World!`.

> ## Prerequisites

Author: I have added the needed deps inside the builder image.

> - A Kubernetes cluster with Knative installed. Follow the

Reviewer: Consider applying this set of edits to make the list parallel and highlight that Knative Build is an optional requirement.

  [installation instructions](https://github.com/knative/docs/blob/master/docs/install/README.md)
  if you need to create one.

> - [Docker](https://www.docker.com) installed and running on your local machine,

Reviewer: revised this to clarify that docker is needed regardless, since you must push your container image to Docker Hub

   FROM fabric8/java-jboss-openjdk8-jdk:1.5.4
   USER jboss
   ENV JAVA_APP_DIR=/deployments

   COPY --from=builder /project/target/lib/* /deployments/lib/
   COPY --from=builder /project/target/*-runner.jar /deployments/app.jar

   ENTRYPOINT [ "/deployments/run-java.sh" ]
   ```

1. Create a new file, `service.yaml` and copy the following service definition
into the file. Make sure to replace `{username}` with your Docker Hub
username.

   ```yaml
   apiVersion: serving.knative.dev/v1alpha1
   kind: Service
   metadata:
     name: helloworld-java-quarkus
   spec:
     runLatest:
       configuration:
         revisionTemplate:
           spec:
             container:
               image: docker.io/{username}/helloworld-java-quarkus
               env:
                 - name: MESSAGE_PREFIX
                   value: "Namaste"
   ```

## Building and deploying the sample

Once you have recreated the sample code files (or used the files in the sample
folder) you're ready to build and deploy the sample app.

We will be using hub.docker.com as the container registry, if you dont have an account yet have one created before proceeding further.

> 

Reviewer: Are these both required?

### Using Docker

> Use Docker to build the sample code into a container. To build and push with

Reviewer: ```suggestion
To use Docker to locally build your container image and and push to Docker Hub, run the following commands with your Docker Hub username: 
```

Reviewer: Add this to clarify that this is optional.

> Docker Hub, run these commands replacing `{username}` with your Docker Hub

Reviewer: ```suggestion
```

> username:

Reviewer: ```suggestion
```

```shell

# We will use the docker daemon of the minikube
eval $(minikube docker-env)

# build the jar and libs
./mvnw clean package

# Build the container on your local machine
docker build -t {username}/helloworld-java-quarkus .

# Push the container to docker registry
docker push {username}/helloworld-java-quarkus
```

After the build has completed and the container is pushed to docker hub, you
can deploy the app into your cluster. Ensure that the container image value
in `service.yaml` matches the container you built in the previous step. Apply
the configuration using `kubectl`:

```shell
kubectl apply --filename service.yaml
```

### Using Knative Build

> The sources also has a knative build file called `quarkus-build.yaml`, that will perform docker less container image build using [buildah](https://buildah.io)

Reviewer: ```suggestion
To use Knative Build, you add your Docker Hub information to the `quarkus-build.yaml` file.
```

Moved optional info about build to the Before you begin section.

Reviewer: There is no mention of installing Knative Build until now, which is late in the flow and forces the user to go back and install the Build component if in fact thats what they prefer.

> ---

Reviewer: Delete lines 1-9 -> the _index.md file "includes" this readme during the build and the frontmatter nor the readfile statement should be here

Reviewer 2: removed

title: "Hello World - Java (Quarkus)"
linkTitle: "Java"
weight: 1
type: "docs"
---

{{% readfile file="README.md" relative="true" markdown="true" %}}

A simple [JAX-RS REST API](https://github.com/jax-rs) application that is written in Java and uses [QuarkusIO](https://quarkus.io/).

It uses an ENV variable `MESSAGE_PREFIX`(defaults to value `Hello`) and prints a message of format `${MESSAGE_PREFIX} Knative World!`.

## Before you begin

> You must meet the following requirements to run this sample:

Reviewer: Please see open comments/suggestions from first review. 
- Is both Knative Build and Docker required?
- The suggested edits make the list parallel and highlight options between docker or Knative Build

- A Kubernetes cluster with Knative installed. Follow the
  [installation instructions](https://github.com/knative/docs/blob/master/docs/install/README.md)
  if you need to create one.

> - [Docker](https://www.docker.com) installed and running on your local machine,

Reviewer: Reordering steps to highlight Knative Build as optional (and that Docker is required to push to Docker Hub):
```suggestion
- You must install [Docker](https://www.docker.com) and have a Docker Hub account to which 
   you are able to upload your sample's container image.
- You can choose to use Knative Build or use Docker to build your sample's container image.
   - If you choose to use the [Knative Build component](../../../../docs/build/installing-build-component.md), 
     you must ensure that the Knative Build component is also installed in the cluster where you installed 
     Knative Serving. This method uses the [buildah](https://buildah.io) tool.
```

Reviewer 2: removed knative build section.

  and a Docker Hub account configured (we'll use it for a container registry).
- You have installed
  [Java SE 8 or later JDK](https://www.eclipse.org/openj9/).

## Recreating the sample code

While you can clone all of the code from this directory, hello world apps are
generally more useful if you build them step-by-step. The following instructions
recreate the source files from this folder.

> 1. From the console, create a new empty web project using the maven archetype

Reviewer: For each step, we should be clear about where you want a file created. Is this at the root? Otherwise users are forced to look at and copy the repo.

Reviewer 2: what does it mean "is this at the root"? current verbiage seems to be inline w/ other examples, e.g., https://github.com/knative/docs/tree/master/docs/serving/samples/hello-world/helloworld-java-spring#recreating-the-sample-code

   commands:

   ```shell
   mvn io.quarkus:quarkus-maven-plugin:0.13.3:create \
    -DprojectGroupId=com.redhat.developer.demos \
    -DprojectArtifactId=helloworld-java-quarkus \
    -DclassName="com.redhat.developer.demos.GreetingResource" \
    -Dpath="/"
   ```

> 1. Update the `GreetingResource` class in

Reviewer: I think we want the user to create this file too? We should first say create that file, state the location, and filename. Then, we have set the context and its clear what class is referred to.

   `src/main/java/com/redhat/developers/demo/GreetingResource.java` with content as:

   ```java
   package com.redhat.developer.demos;

   import javax.inject.Inject;
   import javax.ws.rs.GET;
   import javax.ws.rs.Path;
   import javax.ws.rs.Produces;
   import javax.ws.rs.core.MediaType;
   import org.eclipse.microprofile.config.inject.ConfigProperty;

   @Path("/")
   public class GreeterResource {

     @Inject
     GreetingService greetingService;

     @ConfigProperty(name = "message.prefix")
     String messagePrefix;

     @GET
     @Produces(MediaType.TEXT_PLAIN)
     @Path("/")
     public String greet() {
       return greetingService.greet(messagePrefix);
     }

     @GET
     @Produces(MediaType.TEXT_PLAIN)
     @Path("/healthz")
     public String health() {
       return "OK";
     }
   }

   ```

> 1. Add a new class the `GreetingService` class in

Reviewer: Same comment as above.

         <artifactId>quarkus-resteasy</artifactId>
       </dependency>
       <dependency>
         <groupId>io.quarkus</groupId>
         <artifactId>quarkus-junit5</artifactId>
         <scope>test</scope>
       </dependency>
       <dependency>
         <groupId>io.rest-assured</groupId>
         <artifactId>rest-assured</artifactId>
         <scope>test</scope>
       </dependency>
     </dependencies>
     <profiles>
       <profile>
         <id>native</id>
         <build>
           <plugins>
             <plugin>
               <groupId>io.quarkus</groupId>
               <artifactId>quarkus-maven-plugin</artifactId>
               <version>${quarkus.version}</version>
               <executions>
                 <execution>
                   <goals>
                     <goal>native-image</goal>
                   </goals>
                   <configuration>
                     <enableHttpUrlHandler>true</enableHttpUrlHandler>
                   </configuration>
                 </execution>
               </executions>
             </plugin>
           </plugins>
         </build>
       </profile>
     </profiles>
   </project>
   ```

> 1. Run the application locally:

Reviewer: This should be a new section because its no longer about "recreating the code". Maybe "Locally testing your sample"

Reviewer 2: this is also inline w/ other samples w/ precedence. https://github.com/knative/docs/tree/master/docs/serving/samples/hello-world/helloworld-java-spring#recreating-the-sample-code

   ```shell
   ./mvnw compile quarkus:dev
   ```

   Go to `http://localhost:8080/` to see your `Hello World!` message.

> 1. In your project directory, create a file named `Dockerfile` and copy the code

Reviewer: Where in the project directory? At root?

> 1. In your project directory, create a file named `Dockerfile` and copy the code

Reviewer: Should all of these steps be moved before "locally testing your sample"? We should organize the content into distinct steps/sections so that its clear what the goal/purpose is at each stage in the guide.

Reviewer: If the user chooses Knative Build, what would those steps be?

Reviewer 2: removed knative build section. that should prob go into knative build-template separately.

   block below into it.

   ```docker
   FROM quay.io/rhdevelopers/quarkus-java-builder:graal-1.0.0-rc15 as builder
   COPY . /project
   WORKDIR /project
   # uncomment this to set the MAVEN_MIRROR_URL of your choice, to make faster builds
   # ARG MAVEN_MIRROR_URL=<your-maven-mirror-url>
   # e.g.
   #ARG MAVEN_MIRROR_URL=http://192.168.64.1:8081/nexus/content/groups/public

   RUN /usr/local/bin/entrypoint-run.sh mvn -DskipTests clean package

   FROM fabric8/java-jboss-openjdk8-jdk:1.5.4
   USER jboss
   ENV JAVA_APP_DIR=/deployments

   COPY --from=builder /project/target/lib/* /deployments/lib/
   COPY --from=builder /project/target/*-runner.jar /deployments/app.jar

   ENTRYPOINT [ "/deployments/run-java.sh" ]
   ```

>    If you want to build Quarkus native image, then copy the following code block in to file called `Dockerfile.native`

Reviewer: This is not a standalone element in the list and starts with "If", does that mean its optional?
